### PR TITLE
adding a simple convnet test in interpreter unit test

### PR DESCRIPTION
**Pull request by beicy**
two classes of images : lines and crosses. Current status: the verification test passes for all images used in training set. But for the ones not in training set, get the exactly opposite verification result. Please provide any comments. (Btw, tried to use pytorch to test the convnet first, but had problem with using Pytorch. Still working on it)

### CODE REVIEW

File: tests/unittests/InterpreterTest.cpp [hunk 1]
```diff
@@ -896,3 +896,94 @@ TEST(Interpreter, nonLinearClassifier) {
...
+                                   VisibilityKind::Public);
+
+  auto *ex = mod.createVariable(ElemKind::IndexTy, {numSamples, 1}, "exp",
+                                VisibilityKind::Public, Variable::TrainKind::None);
+  auto *CV0 = F->createConv("conv1", input, 16, 5, 1, 2, 1);
+  auto *RL0 = F->createRELU("relu1", CV0);
+  auto *MP0 = F->createPoolMax("pool1", RL0, 2, 2, 0);
+
+  auto *FCL1 = F->createFullyConnected("fc", MP0, 10);
+  auto *FCL2 = F->createFullyConnected("fc", FCL1, 2);
```
Comment (bertmaher) [inline tests/unittests/InterpreterTest.cpp]: You should probably either get rid of one of these FCs, or insert a nonlinearity (ReLU, maybe?) between them.  Since FCs are just matrix multiply, two in a row can be done with a single one.

Comment (beicy) [reply tests/unittests/InterpreterTest.cpp]: did that before, but same testing result.

File: tests/unittests/InterpreterTest.cpp [hunk 2]
```diff
@@ -896,3 +896,91 @@ TEST(Interpreter, nonLinearClassifier) {
...
+  
+  Function *TF = glow::differentiate(F, EE.getConfig());
+  EE.compile(CompilationMode::Train, TF);
+
+  Tensor images(ElemKind::FloatTy, {numSamples, 32, 32, 1});
+  Tensor labels(ElemKind::IndexTy, {numSamples, 1});
+  generateImageData(images, labels, numSamples);
+
+  // Training:
+  EE.runBatch(150, {input, ex}, {&images, &labels});
```
Comment (artemrakhov) [inline tests/unittests/InterpreterTest.cpp]: Could you please try more iterations and check if rate goes up?

File: tests/unittests/InterpreterTest.cpp [hunk 3]
```diff
@@ -896,3 +896,91 @@ TEST(Interpreter, nonLinearClassifier) {
...
+
+  EE.getConfig().learningRate = 0.001;
+  EE.getConfig().momentum = 0.9;
+
+  auto &mod = EE.getModule();
+  Function *F = mod.createFunction("main");
+  F->setName("simpleConv");
+
+  auto *input = mod.createVariable(ElemKind::FloatTy, {numSamples, 32, 32, 1}, "input",
+                                   VisibilityKind::Public);
```
Comment (artemrakhov) [inline tests/unittests/InterpreterTest.cpp]: Please add `Variable::TrainKind::None`
You don't want to "train" the input.

File: tests/unittests/InterpreterTest.cpp (hunk 2 above)
Comment (beicy) [reply tests/unittests/InterpreterTest.cpp]: Increased to 500, but the rate isn't change

**Comment by beicy** [comment]
For the newly uploaded version, I modified the convnet architecture and generated test cases randomly, now the success rate is 82%.

File: tests/unittests/InterpreterTest.cpp [hunk 4]
```diff
@@ -896,3 +896,84 @@ TEST(Interpreter, nonLinearClassifier) {
...
+
+/// Generate images in two classes. A "line" is labeled as 0
+/// and a "cross" is labeled as 1
+void generateImageData(Tensor &images, Tensor &labels, unsigned numSamples) {
+  auto L = labels.getHandle<size_t>();
+  auto Im = images.getHandle<>();
+  images.zero();
+
+  for (size_t i = 0; i < numSamples; i++) {
+    L.at({i, 0}) = i % 2;
```
Comment (nadavrot) [inline tests/unittests/InterpreterTest.cpp]: One way to make this code more readable is to write:

bool isLine = i % 2; 


L.at({i, 0}) = isLine; 


 if (isLine) {
...

File: tests/unittests/InterpreterTest.cpp [hunk 5]
```diff
@@ -896,3 +896,84 @@ TEST(Interpreter, nonLinearClassifier) {
...
+  Function *F = mod.createFunction("main");
+  F->setName("simpleConv");
+
+  auto *input = mod.createVariable(ElemKind::FloatTy, {numSamples, 32, 32, 1}, "input",
+                                   VisibilityKind::Public, Variable::TrainKind::None);
+
+  auto *ex = mod.createVariable(ElemKind::IndexTy, {numSamples, 1}, "exp",
+                                VisibilityKind::Public, Variable::TrainKind::None);
+
+  auto *CV0 = F->createConv("conv1", input, 6, 5, 1, 2, 1);
```
Comment (nadavrot) [inline tests/unittests/InterpreterTest.cpp]: You should be able to learn this network with a single Conv and a single FC.

File: tests/unittests/InterpreterTest.cpp [hunk 6]
```diff
@@ -896,3 +896,84 @@ TEST(Interpreter, nonLinearClassifier) {
...
+  auto *AP0 = F->createPoolAvg("pool1", CV0, 2, 2, 0);
+  auto *CV1 = F->createConv("conv2", AP0, 6, 5, 1, 2, 1);
+  auto *AP1 = F->createPoolAvg("pool2", CV1, 2, 2, 0);
+  auto *TANH1 = F->createTanh("tanh2", AP1);  
+  auto *FCL1 = F->createFullyConnected("fc", TANH1, 120);
+  auto *TANH2 = F->createTanh("tanh3", FCL1);
+  auto *FCL2 = F->createFullyConnected("fc1", TANH2, 2);
+  auto *SM = F->createSoftMax("sm", FCL2, ex);
```
Comment (nadavrot) [inline tests/unittests/InterpreterTest.cpp]: How many labels are you generating here?

Can you use F->dumpDAG() and check the shape of the output?

File: tests/unittests/InterpreterTest.cpp [hunk 7]
```diff
@@ -896,3 +896,84 @@ TEST(Interpreter, nonLinearClassifier) {
...
+  
+  Function *TF = glow::differentiate(F, EE.getConfig());
+  EE.compile(CompilationMode::Train, TF);
+
+  Tensor images(ElemKind::FloatTy, {numSamples, 32, 32, 1});
+  Tensor labels(ElemKind::IndexTy, {numSamples, 1});
+  generateImageData(images, labels, numSamples);
+
+  // Training:
+  EE.runBatch(150, {input, ex}, {&images, &labels});
```
Comment (nadavrot) [inline tests/unittests/InterpreterTest.cpp]: 150 iterations looks low for your learning rate. You can'e even move from zero to one with your learning rate.

File: tests/unittests/InterpreterTest.cpp [hunk 8]
```diff
@@ -896,3 +896,84 @@ TEST(Interpreter, nonLinearClassifier) {
...
+  /// generate the images used for testing.
+  Tensor test_im(ElemKind::FloatTy, {numSamples, 32, 32, 1});                                                            
+  generateImageData(test_im, labels, numSamples);
+  EE.run({input}, {&test_im});
+  int success = 0;
+  auto SMH = result->getVariable()->getPayload().getHandle<>();
+  for (size_t i = 0; i < numSamples; i++) {                                                                                                                     
+    auto A = SMH.at({i, 0});
+    auto B = SMH.at({i, 1});
+    if ((i % 2 == 0 && A > B) || (i % 2 == 1 && B > A))
```
Comment (nadavrot) [inline tests/unittests/InterpreterTest.cpp]: This code is not readable.

**Comment by rdzhabarov** [comment]
can you also have [interpreter] as part of the name of the PR. See, here: https://github.com/pytorch/glow/blob/master/CONTRIBUTING.md#pull-requests

File: tests/unittests/InterpreterTest.cpp [hunk 9]
```diff
@@ -896,3 +896,84 @@ TEST(Interpreter, nonLinearClassifier) {
     EXPECT_NEAR(RH.at({0, std::get<2>(tests[i])}), 1.0, 0.2);
   }
 }
+
+/// Generate images in two classes. A "line" is labeled as 0
+/// and a "cross" is labeled as 1
+void generateImageData(Tensor &images, Tensor &labels, unsigned numSamples) {
```
Comment (rdzhabarov) [inline tests/unittests/InterpreterTest.cpp]: static method or in private namespace.

File: tests/unittests/InterpreterTest.cpp [hunk 10]
```diff
@@ -896,3 +896,84 @@ TEST(Interpreter, nonLinearClassifier) {
...
+  auto L = labels.getHandle<size_t>();
+  auto Im = images.getHandle<>();
```
Comment (rdzhabarov) [inline tests/unittests/InterpreterTest.cpp]: s/Im/image

File: tests/unittests/InterpreterTest.cpp [hunk 11]
```diff
@@ -896,3 +896,84 @@ TEST(Interpreter, nonLinearClassifier) {
...
+  images.zero();
+
+  for (size_t i = 0; i < numSamples; i++) {
+    L.at({i, 0}) = i % 2;
+    size_t target = rand() % 30 + 1;
+    for (size_t x = 0; x < 32; x++) {
+      for (size_t y = 0; y < 32; y++) {
+        if (i % 2 == 0) {
+          /// line
```
Comment (rdzhabarov) [inline tests/unittests/InterpreterTest.cpp]: /// is goxygen comments. // should be fine here.

File: tests/unittests/InterpreterTest.cpp [hunk 12]
```diff
@@ -896,3 +896,84 @@ TEST(Interpreter, nonLinearClassifier) {
...
+            Im.at({i, x, y, 0u}) = 1;
+        } else {
+          /// cross
+          if (x == target || y == target)
+            Im.at({i, x, y, 0u}) = 1;
+        }
+      }
+    }
+  }
+  printf("\n");
```
Comment (rdzhabarov) [inline tests/unittests/InterpreterTest.cpp]: remove.

File: tests/unittests/InterpreterTest.cpp [hunk 13]
```diff
@@ -896,3 +896,84 @@ TEST(Interpreter, nonLinearClassifier) {
...
+TEST(Interpreter, simpleConv) {
+  unsigned numSamples = 50;
+  ExecutionEngine EE;
+
+  EE.getConfig().learningRate = 0.001;
+  EE.getConfig().momentum = 0.9;
+
+  auto &mod = EE.getModule();
+  Function *F = mod.createFunction("main");
+  F->setName("simpleConv");
```
Comment (nadavrot) [inline tests/unittests/InterpreterTest.cpp]: You don't need to set the name here. The function is created with a name.

File: tests/unittests/InterpreterTest.cpp [hunk 14]
```diff
@@ -896,3 +896,84 @@ TEST(Interpreter, nonLinearClassifier) {
...
+  EE.run({input}, {&test_im});
+  int success = 0;
+  auto SMH = result->getVariable()->getPayload().getHandle<>();
+  for (size_t i = 0; i < numSamples; i++) {                                                                                                                     
+    auto A = SMH.at({i, 0});
+    auto B = SMH.at({i, 1});
+    if ((i % 2 == 0 && A > B) || (i % 2 == 1 && B > A))
+      success++;
+  }
+  llvm::outs() << "the rate of successful tests in testing set: " << (float)success / (float)numSamples << "\n";
```
Comment (rdzhabarov) [inline tests/unittests/InterpreterTest.cpp]: this needs to be an assertion in a form of EXPECT_ ...

Comment (rdzhabarov) [inline tests/unittests/InterpreterTest.cpp]: single (float) is enough.

File: tests/unittests/InterpreterTest.cpp [hunk 15]
```diff
@@ -896,3 +896,84 @@ TEST(Interpreter, nonLinearClassifier) {
...
+  // Training:
+  EE.runBatch(150, {input, ex}, {&images, &labels});
+  EE.compile(CompilationMode::Infer, F);
+
+  /// generate the images used for testing.
+  Tensor test_im(ElemKind::FloatTy, {numSamples, 32, 32, 1});                                                            
+  generateImageData(test_im, labels, numSamples);
+  EE.run({input}, {&test_im});
+  int success = 0;
+  auto SMH = result->getVariable()->getPayload().getHandle<>();
```
Comment (rdzhabarov) [inline tests/unittests/InterpreterTest.cpp]: iirc, you can result->getVariable()->getHandle() directly.

File: tests/unittests/InterpreterTest.cpp [hunk 16]
```diff
@@ -896,3 +896,84 @@ TEST(Interpreter, nonLinearClassifier) {
...
+  Tensor images(ElemKind::FloatTy, {numSamples, 32, 32, 1});
+  Tensor labels(ElemKind::IndexTy, {numSamples, 1});
+  generateImageData(images, labels, numSamples);
+
+  // Training:
+  EE.runBatch(150, {input, ex}, {&images, &labels});
+  EE.compile(CompilationMode::Infer, F);
+
+  /// generate the images used for testing.
+  Tensor test_im(ElemKind::FloatTy, {numSamples, 32, 32, 1});
```
Comment (rdzhabarov) [inline tests/unittests/InterpreterTest.cpp]: s/test_im/testImage

File: tests/unittests/InterpreterTest.cpp [hunk 17]
```diff
@@ -896,3 +896,84 @@ TEST(Interpreter, nonLinearClassifier) {
...
+
+  Tensor images(ElemKind::FloatTy, {numSamples, 32, 32, 1});
+  Tensor labels(ElemKind::IndexTy, {numSamples, 1});
+  generateImageData(images, labels, numSamples);
+
+  // Training:
+  EE.runBatch(150, {input, ex}, {&images, &labels});
+  EE.compile(CompilationMode::Infer, F);
+
+  /// generate the images used for testing.
```
Comment (rdzhabarov) [inline tests/unittests/InterpreterTest.cpp]: comments are //, /// is for doxygen.

**Comment by artemrakhov** [comment]
I thought more about this task of distinguishing line from cross, and came up with the following net architecture.
```
Conv (3x3 kernel, depthOut=1)
Relu
FC (depthOut=2)
Softmax
```
This net will be able to reach 100% accuracy rate with the following weights:
```
Conv.filter:
1 1 1
1 1 1
1 1 1
Conv.bias:
-4
FC.weight:
0 1
0 1
0 1
...
0 1
FC.bias:
0.5 0
```
How it works: if we have a cross, it means that there is a `1` pixel with 4 neighboring `1` pixels. It means that after applying filter we'll have `5` at such pixels, and something smaller at others. After we apply bias we'll have 1 for such "cross" points, and <= 0 at other pixels. We replace <= 0 with 0 by doing Relu. Now FC will transform image to 2 classes:

1. Column of zeros, bias = 0.5. Don't even look at pixels, just assign 0.5
1. Column of ones, bias = 0. Summarize all pixels. If we have no cross points, sum will be 0. Otherwise it'll be 1 (or number of cross points).

If we have cross point, result will be (0.5, 1), otherwise result will be (0.5, 0).

Now, it's not necessary for the net to converge to the weights that I just described. It may converge to something else. I just wanted to prove that it's possible to reach 100% accuracy with this architecture.
